src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+        }
+
+        void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, string oldMoniker, ITextBuffer textBuffer)
+        {
+        }
+
+        private class OpenSourceGeneratedFile : ForegroundThreadAffinitizedObject, IDisposable
+        {
+            private readonly SourceGeneratedFileManager _fileManager;
+            private readonly ITextBuffer _textBuffer;
+            private readonly Workspace _workspace;
+            private readonly ProjectId _projectId;
+            private readonly string _generatorTypeName;
+            private readonly string _generatedSourceHintName;
+
+            /// <summary>
+            /// A read-only region that we create across the entire file to prevent edits unless we are the one making them.
+            /// It's a dynamic read-only region that will allow edits if <see cref="_updatingBuffer"/> is set.
+            /// </summary>
+            private readonly IReadOnlyRegion _readOnlyRegion;
+            private bool _updatingBuffer = false;
+
+            /// <summary>
+            /// A cancellation token used for any background updating of this file; this is cancelled on the UI thread
+            /// when the file is closed.
+            /// </summary>
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+
+            /// <summary>
+            /// A queue used to batch updates to the file.
+            /// </summary>
+            private readonly AsyncBatchingDelay _batchingWorkQueue;
+
+            /// <summary>
+            /// The <see cref="IVsWindowFrame"/> of the active window. This may be null if we're in the middle of construction and
+            /// we haven't been given it yet.
+            /// </summary>
+            private IVsWindowFrame? _windowFrame;
+
+            private string? _windowFrameMessageToShow = null;
+            private ImageMoniker _windowFrameImageMonikerToShow = default;
+            private string? _currentWindowFrameMessage = null;
+            private ImageMoniker _currentWindowFrameImageMoniker = default;
+            private IVsInfoBarUIElement? _currentWindowFrameInfoBarElement = null;
+
+            public OpenSourceGeneratedFile(SourceGeneratedFileManager fileManager, ITextBuffer textBuffer, Workspace workspace, ProjectId projectId, string generatorTypeName, string generatedSourceHintName, IThreadingContext threadingContext)
+                : base(threadingContext, assertIsForeground: true)
+            {
+                _fileManager = fileManager;
+                _textBuffer = textBuffer;
+                _workspace = workspace;
+                _projectId = projectId;
+                _generatorTypeName = generatorTypeName;
+                _generatedSourceHintName = generatedSourceHintName;
+
+                // We'll create a read-only region for the file, but it'll be a dynamic region we can temporarily suspend
+                // while we're doing edits.
```

Comment:
❔ What API does Metadata As Source use to mark files as read-only?

Reply:
I was trying to find this too, in CPS, because I know I've done it before, but then I remembered it was in a branch that didn't get merged. I _think_ you mark the file read only in the RDT

Reply:
In Metadata as Source's case, we mark the actual file on disk as read-only, and then we end up with the lock icon via this magic:

https://github.com/dotnet/roslyn/blob/b1ee64f2af58fd0a5da7fa343fa649efa999b87d/src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs#L121-L128

I'm very hazy on the details but I think one of those don't actually make the buffer read only at all (i.e. it's a flag that you can read, but if you don't read it nothing blocks from editing), and one I think does create buffer read only spans but was worried about other side effects.

Reply:
Looks like just setting the state flags on the buffer should do it. This is all that the project system does when you open something from the .NET SDK (via the Imports node), and the tab is readonly but the file on disk isn't.

https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/ProjectImports/ImportTreeCommandGroupHandler.cs#L112-L116